sql/core/src/main/scala/org/apache/spark/sql/Observation.scala

```diff
@@ -0,0 +1,189 @@
+  private val listener: ObservationListener = ObservationListener(this)
+
+  private var sparkSession: Option[SparkSession] = None
+
+  @transient private var row: Option[Row] = None
+
+  /**
+   * Attach this observation to the given Dataset.
+   * Remember to call `close()` when the observation is done.
+   *
+   * @param ds dataset
+   * @tparam T dataset type
+   * @return observed dataset
+   */
+  def on[T](ds: Dataset[T])(expr: Column, exprs: Column*): Dataset[T] = {
+    if (ds.isStreaming) {
+      throw new IllegalArgumentException("Observation does not support streaming Datasets")
+    }
+    register(ds.sparkSession)
+    ds.observe(name, expr, exprs: _*)
+  }
+
+  /**
+   * Get the observation results. Waits for the first action on the observed dataset to complete.
+   * After calling `reset()`, waits for completion of the next action on the observed dataset.
+   */
+  def get: Row = option().get
+
+  /**
+   * Get the observation results. Waits for the first action on the observed dataset to complete.
+   * This method times out waiting for the action after the given amount of time.
+   * After calling `reset()`, waits for completion of the next action on the observed dataset.
+   *
+   * @param time timeout
+   * @param unit timeout time unit
+   * @return observation row as an Option, or None on timeout
+   */
+  def option(time: Long, unit: TimeUnit): Option[Row] = option(Some(time), unit)
+
+  /**
+   * Wait for the first action on the observed dataset to complete.
+   * When the time parameter is given, this method times out waiting for the action.
+   * After calling `reset()`, waits for completion of the next action on the observed dataset.
+   *
+   * @param time timeout
+   * @param unit timeout time unit
+   * @return true if action complete within timeout, false on timeout
+   */
+  def waitCompleted(time: Option[Long] = None, unit: TimeUnit = TimeUnit.MILLISECONDS): Boolean = {
+    lock.lock()
+    try {
+      if (row.isEmpty) {
+        if (time.isDefined) {
+          completed.await(time.get, unit)
+        } else {
+          completed.await()
+        }
+      }
+      row.isDefined
+    } finally {
+      lock.unlock()
+    }
+  }
+
+  /**
+   * Wait for the first action on the observed dataset to complete.
+   * After calling `reset()`, waits for completion of the next action on the observed dataset.
+   *
+   * @param time timeout
+   * @param unit timeout time unit
+   * @return true if action complete within timeout, false on timeout
+   */
+  def waitCompleted(time: Long, unit: TimeUnit): Boolean = waitCompleted(Some(time), unit)
+
+  /**
+   * Reset the observation. This deletes the observation and allows to wait for completion
+   * of the next action called on the observed dataset. Not resetting the observation before
+   * attempting to retrieve the next action's results via get, option or waitCompleted is not
+   * guaranteed to work.
+   */
+  def reset(): Unit = {
```

Comment:
I'm a little worried about this API. The listener is registered and the result of the first observation may come at any time later.

Shall we let the `Observation` instance be used only once and not provide `reset` API?

Reply:
Sounds good to me! Why would you re-process the same DataFrame twice, you would try to avoid this in most cases, so the use of `reset` is a bit of an edge case.